// @reportportal/agent-js-playwright/build/constants/index.js
"use strict";
/*
 *  Copyright 2021 EPAM Systems
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BASIC_ATTACHMENT_NAMES = exports.BASIC_ATTACHMENT_CONTENT_TYPES = exports.TEST_OUTCOME_TYPES = exports.TEST_ANNOTATION_TYPES = exports.PREDEFINED_LOG_LEVELS = exports.STATUSES = exports.TEST_ITEM_TYPES = exports.LAUNCH_MODES = void 0;
var launchModes_1 = require("./launchModes");
Object.defineProperty(exports, "LAUNCH_MODES", { enumerable: true, get: function () { return launchModes_1.LAUNCH_MODES; } });
var testItemTypes_1 = require("./testItemTypes");
Object.defineProperty(exports, "TEST_ITEM_TYPES", { enumerable: true, get: function () { return testItemTypes_1.TEST_ITEM_TYPES; } });
var statuses_1 = require("./statuses");
Object.defineProperty(exports, "STATUSES", { enumerable: true, get: function () { return statuses_1.STATUSES; } });
var logLevels_1 = require("./logLevels");
Object.defineProperty(exports, "PREDEFINED_LOG_LEVELS", { enumerable: true, get: function () { return logLevels_1.PREDEFINED_LOG_LEVELS; } });
var playwright_1 = require("./playwright");
Object.defineProperty(exports, "TEST_ANNOTATION_TYPES", { enumerable: true, get: function () { return playwright_1.TEST_ANNOTATION_TYPES; } });
Object.defineProperty(exports, "TEST_OUTCOME_TYPES", { enumerable: true, get: function () { return playwright_1.TEST_OUTCOME_TYPES; } });
Object.defineProperty(exports, "BASIC_ATTACHMENT_CONTENT_TYPES", { enumerable: true, get: function () { return playwright_1.BASIC_ATTACHMENT_CONTENT_TYPES; } });
Object.defineProperty(exports, "BASIC_ATTACHMENT_NAMES", { enumerable: true, get: function () { return playwright_1.BASIC_ATTACHMENT_NAMES; } });
