// @reportportal/agent-js-playwright/build/constants/logLevels.js
"use strict";
/*
 *  Copyright 2021 EPAM Systems
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PREDEFINED_LOG_LEVELS = void 0;
var PREDEFINED_LOG_LEVELS;
(function (PREDEFINED_LOG_LEVELS) {
    PREDEFINED_LOG_LEVELS["TRACE"] = "TRACE";
    PREDEFINED_LOG_LEVELS["DEBUG"] = "DEBUG";
    PREDEFINED_LOG_LEVELS["WARN"] = "WARN";
    PREDEFINED_LOG_LEVELS["INFO"] = "INFO";
    PREDEFINED_LOG_LEVELS["ERROR"] = "ERROR";
    PREDEFINED_LOG_LEVELS["FATAL"] = "FATAL";
})(PREDEFINED_LOG_LEVELS = exports.PREDEFINED_LOG_LEVELS || (exports.PREDEFINED_LOG_LEVELS = {}));
