// @reportportal/agent-js-playwright/build/constants/playwright.js
"use strict";
/*
 *  Copyright 2023 EPAM Systems
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BASIC_ATTACHMENT_NAMES = exports.BASIC_ATTACHMENT_CONTENT_TYPES = exports.TEST_OUTCOME_TYPES = exports.TEST_ANNOTATION_TYPES = void 0;
var TEST_ANNOTATION_TYPES;
(function (TEST_ANNOTATION_TYPES) {
    TEST_ANNOTATION_TYPES["FIXME"] = "fixme";
    TEST_ANNOTATION_TYPES["SKIP"] = "skip";
    TEST_ANNOTATION_TYPES["FAIL"] = "fail";
})(TEST_ANNOTATION_TYPES = exports.TEST_ANNOTATION_TYPES || (exports.TEST_ANNOTATION_TYPES = {}));
var TEST_OUTCOME_TYPES;
(function (TEST_OUTCOME_TYPES) {
    TEST_OUTCOME_TYPES["SKIPPED"] = "skipped";
    TEST_OUTCOME_TYPES["EXPECTED"] = "expected";
    TEST_OUTCOME_TYPES["UNEXPECTED"] = "unexpected";
    TEST_OUTCOME_TYPES["FLAKY"] = "flaky";
})(TEST_OUTCOME_TYPES = exports.TEST_OUTCOME_TYPES || (exports.TEST_OUTCOME_TYPES = {}));
var BASIC_ATTACHMENT_CONTENT_TYPES;
(function (BASIC_ATTACHMENT_CONTENT_TYPES) {
    BASIC_ATTACHMENT_CONTENT_TYPES["VIDEO"] = "video/webm";
    BASIC_ATTACHMENT_CONTENT_TYPES["TRACE"] = "application/zip";
})(BASIC_ATTACHMENT_CONTENT_TYPES = exports.BASIC_ATTACHMENT_CONTENT_TYPES || (exports.BASIC_ATTACHMENT_CONTENT_TYPES = {}));
var BASIC_ATTACHMENT_NAMES;
(function (BASIC_ATTACHMENT_NAMES) {
    BASIC_ATTACHMENT_NAMES["VIDEO"] = "video";
    BASIC_ATTACHMENT_NAMES["TRACE"] = "trace";
})(BASIC_ATTACHMENT_NAMES = exports.BASIC_ATTACHMENT_NAMES || (exports.BASIC_ATTACHMENT_NAMES = {}));
